--- bios.asm	2024-07-22 15:26:36.632908086 +0900
+++ new/bios.asm	2024-07-23 12:27:29.581226376 +0900
@@ -2,6 +2,9 @@
 ;
 ;	Copyright (C) 1988-2007 by Udo Munk
 ;
+;       modified for TangNanoZ80MEM-CPM
+;       by Ryo Mukai 
+;
 MSIZE	EQU	64		;cp/m version memory size in kilobytes
 ;
 ;	"bias" is address offset from 3400H for memory systems
@@ -17,8 +20,10 @@
 ;
 ;	I/O ports
 ;
-CONSTA	EQU	0		;console status port
-CONDAT	EQU	1		;console data port
+; CONSTA	EQU	0		;console status port
+; CONDAT	EQU	1		;console data port
+CONDAT	EQU	0		;console data port
+CONSTA	EQU	1		;console status port
 PRTSTA	EQU	2		;printer status port
 PRTDAT	EQU	3		;printer data port
 AUXDAT	EQU	5		;auxiliary data port
@@ -126,8 +131,13 @@
 ;
 ;	messages
 ;
-SIGNON: DEFM	'64K CP/M Vers. 2.2 (Z80 CBIOS V1.2 for Z80SIM, '
+SIGNON:	DEFB	13,10
+	DEFM	'64K CP/M Vers. 2.2 '
+	DEFB	13,10
+	DEFM    '(Z80 CBIOS V1.2 for Z80SIM, '
 	DEFM	'Copyright 1988-2007 by Udo Munk)'
+	DEFB	13,10
+	DEFM	'(modified for TangNanoZ80MEM by Ryo Mukai)'
 	DEFB	13,10,0
 ;
 LDERR:	DEFM	'BIOS: error booting'
@@ -233,18 +243,31 @@
 ;
 ;	console status, return 0ffh if character ready, 00h if not
 ;
-CONST:	IN	A,(CONSTA)	;get console status
+CONST:	IN 	A,(CONSTA)
+	AND 	01H
+	JP 	Z,CONST_NOTRDY
+	LD	A,0FFH
+	RET
+CONST_NOTRDY:
+	LD	A,00H
 	RET
 ;
 ;	console character into register a
 ;
-CONIN:	IN	A,(CONDAT)	;get character from console
+CONIN:	IN 	A,(CONSTA)
+	AND 	01H
+	JP 	Z,CONIN
+	IN 	A,(CONDAT)
+	AND	7FH
 	RET
 ;
 ;	console character output from register c
 ;
-CONOUT: LD	A,C		;get to accumulator
-	OUT	(CONDAT),A	;send character to console
+CONOUT:	IN	A,(CONSTA)
+	AND	04H
+	JP	Z,CONOUT
+	LD	A,C
+	OUT	(CONDAT),A
 	RET
 ;
 ;	list character from register c
